src/state.rs

```diff
@@ -109,6 +113,178 @@ pub struct ResolvedDns {
     dns_refresh_rate: std::time::Duration,
 }
 
+impl DnsResolver {
+    fn new(dns_resolver_cfg: ResolverConfig, dns_resolver_opts: ResolverOpts) -> Self {
+        let resolved_dns = Arc::new(RwLock::new(ResolvedDnsStore::default()));
+        Self {
+            resolved_dns,
+            dns_resolver_cfg,
+            dns_resolver_opts,
+        }
+    }
+
+    fn read(&self) -> RwLockReadGuard<'_, ResolvedDnsStore> {
+        self.resolved_dns.read().unwrap()
+    }
+
+    fn write(&self) -> RwLockWriteGuard<'_, ResolvedDnsStore> {
+        self.resolved_dns.write().unwrap()
+    }
+
+    async fn resolve_host(
+        &mut self,
+        hostname: &String,
+        workload: &Workload,
+    ) -> Option<ResolvedDns> {
+        // optimize so that if multiple requests to the same hostname come in at the same time,
+        // we don't start more than one background on-demand DNS task
+        let is_the_first_req = self.get_singleflight_by_hostname(hostname.to_owned());
```

Comment:
Hi @nmittler, there are 2 mutable borrows occurs in both `self.get_or_create_notify` (first mutable borrow) and `self.resolve_on_demand_dns` (second mutable borrow), so I have to use this `let element = self.get_cached_resolve_dns_for_hostname(hostname);` to avoid this problem. 
However, I can change `get_singleflight_by_hostname` into `get_or_create_notify`.  Does it make sense?